Kernel/ACPI/ACPIStaticParser.cpp

```diff
@@ -104,26 +105,97 @@ namespace ACPI {
 
     bool StaticParser::can_reboot()
     {
-        // FIXME: Determine if we need to do MMIO/PCI/IO access to reboot, according to ACPI spec 6.2, Section 4.8.3.6
         auto region = MM.allocate_kernel_region(m_fadt.page_base(), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
         auto* fadt = (const Structures::FADT*)region->vaddr().offset(m_fadt.offset_in_page().get()).as_ptr();
-        return fadt->h.revision >= 2;
+        if (fadt->h.revision < 2)
+            return false;
+        return (fadt->flags & (u32)FADTFeatureFlags::RESET_REG_SUPPORTED) != 0;
     }
 
     void StaticParser::try_acpi_reboot()
     {
-        // FIXME: Determine if we need to do MMIO/PCI/IO access to reboot, according to ACPI spec 6.2, Section 4.8.3.6
+        InterruptDisabler disabler;
+        if (!can_reboot()) {
+            klog() << "ACPI: Reboot, Not supported!";
+            return;
+        }
 #ifdef ACPI_DEBUG
         dbg() << "ACPI: Rebooting, Probing FADT (" << m_fadt << ")";
 #endif
 
         auto region = MM.allocate_kernel_region(m_fadt.page_base(), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
-        auto* fadt = (const Structures::FADT*)region->vaddr().offset(m_fadt.offset_in_page().get()).as_ptr();
-        if (fadt->h.revision >= 2) {
+        auto* fadt = (const volatile Structures::FADT*)region->vaddr().offset(m_fadt.offset_in_page().get()).as_ptr();
+        switch (fadt->reset_reg.address_space) {
+        case (u8)GenericAddressStructure::AddressSpace::SystemIO:
             klog() << "ACPI: Reboot, Sending value 0x" << String::format("%x", fadt->reset_value) << " to Port 0x" << String::format("%x", fadt->reset_reg.address);
             IO::out8(fadt->reset_reg.address, fadt->reset_value);
+            break;
+        case (u8)GenericAddressStructure::AddressSpace::SystemMemory: {
+            auto p_reset_reg = PhysicalAddress(fadt->reset_reg.address);
+            auto reset_region = MM.allocate_kernel_region(p_reset_reg.page_base(), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
+            klog() << "ACPI: Reboot, Sending value 0x" << String::format("%x", fadt->reset_value) << " to " << p_reset_reg;
+            switch (fadt->reset_reg.access_size) {
+            case (u8)GenericAddressStructure::AccessSize::Byte: {
+                auto* reset_reg = (volatile u8*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
+                (*reset_reg) = fadt->reset_value;
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::Word: {
+                auto* reset_reg = (volatile u16*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
```

Comment:
This is a bit awkward. Or rather, `PhysicalAddress::offset_in_page()` is awkward, and it makes this look awkward.
Let's make `offset_in_page()` return a primitive like `FlatPtr` rather than a `PhysicalAddress`. Then we don't have to call `.get()` on it.
The offset itself is not meaningfully a `PhysicalAddress` anyway. :)